Eagerly emit getters at Onone.

augusto2112 opened the pull request:
Force SILGen to also eagerly emit getters when compiling at Onone. The reason for this is that getters (even not user-written ones, generated by result builders) can, and are often called by users debugging swift programs, and should be available for that reason.

rdar://133329303

augusto2112 commented:
@swift-ci smoke test

slavapestov commented on lib/SIL/IR/SILDeclRef.cpp:
```diff
@@ -387,6 +387,12 @@ bool SILDeclRef::hasUserWrittenCode() const {
 
+bool SILDeclRef::isGetter() const {
+  if (auto *accessor = dyn_cast_or_null<AccessorDecl>(getFuncDecl()))
+    return accessor->isGetter();
```
Is this also going to be true for getters of ClangImporter-synthesized properties? We probably don't want to emit all of those.

augusto2112 replied:
@slavapestov are those ObjC properties that are imported into Swift? Couldn't they also be accessed in the debugger?

slavapestov replied:
When evaluating an objective-C method call, the compiler generates the right thunks with shared linkage. That's why we don't want to force emission of those symbols if they're not needed. So perhaps the mayDelay computation needs to check the SIL linkage.

slavapestov commented on lib/SILGen/SILGen.cpp:
```diff
@@ -1197,9 +1197,16 @@ void SILGenModule::emitOrDelayFunction(SILDeclRef constant) {
   // Implicit decls may be delayed if they can't be used externally.
   auto linkage = constant.getLinkage(ForDefinition);
-  bool mayDelay = !constant.hasUserWrittenCode() &&
+  bool mayDelay = !isOnoneGetter && !constant.hasUserWrittenCode() &&
```
Maybe hasUserWrittenCode() should also return true for property wrappers?

augusto2112 replied:
I thought about that too, since @hamishknight wrote it maybe they can chime in if that'd be appropriate or not.

hamishknight replied:
The intention behind `hasUserWrittenCode` was "does this decl contain any executable code that appears in a source file?", which is needed for code coverage since we don't want to emit coverage maps for anything that won't have any source regions. It's also used here because that check is also what we need for "do we need to emit this for SILOptimizer diagnostics?". It seems like we need a different check here for the debugger which is more "can this decl be referenced from user-written code?". I wonder if we can leverage the `isSynthesized` Decl flag for that, IIRC that's effectively what it models for implicit decls.

hamishknight replied:
Out of curiosity, should this behavior also be limited to cases where we're emitting debug info?

augusto2112 replied:
> Out of curiosity, should this behavior also be limited to cases where we're emitting debug info?

We usually avoid debug info interfering with code generation.

augusto2112 committed:
Eagerly emit getters at Onone.

Force SILGen to also eagerly emit getters when compiling at Onone.
The reason for this is that getters (even not user-written ones,
generated by result builders) can, and are often called by users
debugging swift programs, and should be available for that reason.

rdar://133329303

augusto2112 commented:
@swift-ci smoke test
